tasks/libs/copyright.py

```diff
@@ -0,0 +1,184 @@
+        os.chdir(current_dir)
+
+        return PurePosixPath(repo_dir)
+
+    @staticmethod
+    def _get_matching_files(root_dir, glob_pattern, exclude=None):
+        if exclude is None:
+            exclude = []
+
+        all_matching_files = Path(root_dir).glob(glob_pattern)
+
+        # `all_matching_files` is a generator so we have to do the counting ourselves
+        all_matching_files_cnt = 0
+
+        filtered_files = []
+        for filepath in all_matching_files:
+            all_matching_files_cnt += 1
+
+            excluded = False
+            for matcher in exclude:
+                if re.search(matcher, filepath.as_posix()):
+                    excluded = True
+                    break
+
+            if excluded:
+                continue
+
+            filtered_files.append(filepath)
+
+        excluded_files_cnt = all_matching_files_cnt - len(filtered_files)
+        print(f"[WARN] Excluding {excluded_files_cnt} files based on path filters!")
+
+        return sorted(filtered_files)
+
+    @staticmethod
+    def _get_header(filepath):
+        header = []
+        with open(filepath, "r") as file_obj:
+            # We expect a specific header format which should be 4 lines
+            for _ in range(4):
+                header.append(file_obj.readline().strip())
+
+        return header
+
+    @staticmethod
+    def _is_excluded_header(header, exclude=None):
+        if exclude is None:
+            exclude = []
+
+        for matcher in exclude:
+            if re.search(matcher, header[0]):
+                return True
+
+        return False
+
+    @staticmethod
+    def _has_copyright(filepath, debug=False):
+        header = CopyrightLinter._get_header(filepath)
+        if header is None:
+            print("Mismatch found! Could not find any content in file!")
+            return False
+
+        if len(header) > 0 and CopyrightLinter._is_excluded_header(header, exclude=COMPILED_HEADER_EXCLUSION_REGEX):
+            if debug:
+                print(f"[WARN] Excluding {filepath} based on header '{header[0]}'")
+            return True
+
+        if len(header) <= 3:
+            print("Mismatch found! File too small for header stanza!")
+            return False
+
+        for line_idx, matcher in enumerate(COMPILED_COPYRIGHT_REGEX):
+            if not re.match(matcher, header[line_idx]):
+                print(f"Mismatch found! Expected '{COPYRIGHT_REGEX[line_idx]}' pattern but got '{header[line_idx]}'")
+                return False
+
+        return True
+
+    @staticmethod
+    def _assert_copyrights(files, debug=False):
+        failing_files = []
+        for filepath in files:
+            if CopyrightLinter._has_copyright(filepath, debug=debug):
+                if debug:
+                    print(f"[ OK ] {filepath}")
+
+                continue
+
+            print(f"[FAIL] {filepath}")
+            failing_files.append(filepath)
+
+        total_files = len(files)
+        if failing_files != 0:
```

Comment:
failing_files is a list, maybe you wanted to do this?
```suggestion
        if failing_files:
```

Reply:
Yup - good catch!